hashintel/hash
# Generate OpenAPI Client for the HASH Graph

## 🌟 What is the purpose of this PR?



This PR adds a TypeScript client generated from our Graph OpenAPI Spec. 
The client is generated with https://github.com/openapitools/openapi-generator-cli using `yarn` dependencies.

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1200211978612931/1202705536455591/f) _(internal)_

## 🚫 Blocked by





## 🔍 What does this change?




- Adds a new Makefile.toml CI step to generate OpenAPI client(s)
   - This step can be considered a validation step as well, and will fail if spec is invalid, and also fail if the `cargo make generate-openapi-client` command hasn't been run before merging the PR.
- Check in the TypeScript OpenAPI client, and add it to the Yarn workspace
- Add a dependency on the Graph client from the `hash-api` package
- Make JSON Schemas compatible with the type generator

## 📜 Does this require a change to the docs?



- Added instructions to generating the client.
- Better documentation for the client should be introduced.

## ⚠️ Known issues


- Because the OpenAPI spec is served with the binary, the binary needs to be running before we can generate the client. This is why the CI approach was chosen instead of having pre-commit hooks or equivalent. 

## 🐾 Next steps


- Start migrating the HASH backend!

## 🛡 What tests cover this?



- None so far, but CI makes sure that the client is in line with the OpenAPI spec.

## ❓ How to test this?



1.  Checkout the branch / view the deployment
1.  run `cargo make generate-openapi-client` from the `hash_graph` folder and observe no changes in the git diff
1. make a change in the routes or anything in the spec
1. re-generate client and observe changes.

## 📹 Demo


Not much to screenshot, the `/packages/graph/hash_graph_clients/typescript` folder contains the client.
